extensions/kafka/src/main/java/io/deephaven/kafka/KeyOrValueSpecObjectProcessorImpl.java

```diff
@@ -0,0 +1,129 @@
+        }
+        if (columnNames.stream().distinct().count() != columnNames.size()) {
+            throw new IllegalArgumentException("Expected columnNames to have distinct values");
+        }
+        this.deserializer = Objects.requireNonNull(deserializer);
+        this.processor = Objects.requireNonNull(processor);
+        this.columnNames = List.copyOf(columnNames);
+    }
+
+    @Override
+    public Optional<SchemaProvider> getSchemaProvider() {
+        return Optional.empty();
+    }
+
+    @Override
+    protected Deserializer<? extends T> getDeserializer(KeyOrValue keyOrValue,
+            SchemaRegistryClient schemaRegistryClient,
+            Map<String, ?> configs) {
+        return deserializer;
+    }
+
+    @Override
+    protected KeyOrValueIngestData getIngestData(KeyOrValue keyOrValue, SchemaRegistryClient schemaRegistryClient,
+            Map<String, ?> configs, MutableInt nextColumnIndexMut, List<ColumnDefinition<?>> columnDefinitionsOut) {
+        final KeyOrValueIngestData data = new KeyOrValueIngestData();
+        data.fieldPathToColumnName = new LinkedHashMap<>();
+        final int L = columnNames.size();
+        for (int i = 0; i < L; ++i) {
+            final String columnName = columnNames.get(i);
+            final Type<?> type = processor.outputTypes().get(i);
+            data.fieldPathToColumnName.put(columnName, columnName);
+            columnDefinitionsOut.add(ColumnDefinition.of(columnName, type));
+        }
+        return data;
+    }
+
+    @Override
+    protected KeyOrValueProcessor getProcessor(TableDefinition tableDef, KeyOrValueIngestData data) {
+        return new KeyOrValueProcessorImpl(
+                offsetsFunction(MultiFieldChunkAdapter.chunkOffsets(tableDef, data.fieldPathToColumnName)));
+    }
+
+    private class KeyOrValueProcessorImpl implements KeyOrValueProcessor {
+        private final Function<WritableChunk<?>[], List<WritableChunk<?>>> offsetsAdapter;
+
+        private KeyOrValueProcessorImpl(Function<WritableChunk<?>[], List<WritableChunk<?>>> offsetsAdapter) {
+            this.offsetsAdapter = Objects.requireNonNull(offsetsAdapter);
+        }
+
+        @Override
+        public void handleChunk(ObjectChunk<Object, Values> inputChunk, WritableChunk<Values>[] publisherChunks) {
+            // noinspection unchecked
+            final ObjectChunk<T, ?> in = (ObjectChunk<T, ?>) inputChunk;
+            // we except isInOrder to be true, so apply should be an O(1) op no matter how many columns there are.
+            processor.processAll(in, offsetsAdapter.apply(publisherChunks));
+        }
+    }
+
+    private static <T> Function<T[], List<T>> offsetsFunction(int[] offsets) {
+        return offsets.length == 0
+                ? array -> Collections.emptyList()
+                : isInOrder(offsets)
+                        ? array -> Arrays.asList(array).subList(offsets[0], offsets[0] + offsets.length)
```

Comment:
I dislike this allocation. We could just advise of which `out` chunk index to start from.

Reply:
I'm very much against this idea. My hope is that `ObjectProcessor#processAll` is very straightforward. Adding indexes as part of that feels like a cop-out based on our current implementation. I very much expect that a re-write of some internal implementation details (KafkaStreamPublisher) and continuing along w/ our path of making KafkaTools better will make this whole adapter layer unnecessary in the future.